client/metalink.c

```diff
@@ -0,0 +1,360 @@
+    
+    //Get Hash Content.
+    xmlContValue = xmlNodeGetContent(node);
+    if(xmlContValue == NULL)
+    {
+        dwError = ERROR_TDNF_MLK_PARSER_MISSING_HASH_CONTENT;
+        pr_err("XML Parser Error:HASH value is not present in HASH element", value);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    value = (const char*)xmlContValue;
+    ml_hash_info->value = strdup(value);
+
+cleanup:
+    if(xmlPropValue != NULL)
+    {
+        xmlFree(xmlPropValue);
+        xmlPropValue = NULL;
+    }
+    if(xmlContValue != NULL)
+    {
+        xmlFree(xmlContValue);
+        xmlContValue = NULL;
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFParseUrlTag(
+    TDNF_METALINK_CTX *ml_ctx,
+    xmlNode *node
+    )
+{
+    uint32_t dwError = 0;
+    xmlChar* xmlPropValue = NULL;
+    xmlChar* xmlContValue = NULL;
+    const char *value = NULL;
+    long long prefValue = 0;
+    char *nptr = NULL;
+    TDNF_METALINK_URL_INFO *ml_url_info = NULL;
+    
+    dwError = TDNFAllocateMemory(1, sizeof(TDNF_METALINK_URL_INFO),
+                                 (void**)&ml_url_info);
+    BAIL_ON_TDNF_ERROR(dwError);
+    TDNFAppendList(&(ml_ctx->urls), ml_url_info);
+    
+    if ((xmlPropValue = xmlGetProp(node, ATTR_PROTOCOL)))
+    {
+        value = (const char*)xmlPropValue;
+        ml_url_info->protocol = strdup(value);
+        xmlFree(xmlPropValue);
+        xmlPropValue = NULL;
+        value = NULL;
+    }
+    if ((xmlPropValue = xmlGetProp(node, ATTR_TYPE)))
+    {
+        value = (const char*)xmlPropValue;
+        ml_url_info->type = strdup(value);
+        xmlFree(xmlPropValue);
+        xmlPropValue = NULL;
+        value = NULL;
+    }
+    if ((xmlPropValue = xmlGetProp(node, ATTR_LOCATION)))
+    {
+        value = (const char*)xmlPropValue;
+        ml_url_info->location = strdup(value);
+        xmlFree(xmlPropValue);
+        xmlPropValue = NULL;
+        value = NULL;
+    }
+    if ((xmlPropValue = xmlGetProp(node, ATTR_PREFERENCE)))
+    {
+        value = (const char*)xmlPropValue;
+        prefValue = strtoll(value, &nptr, 0);
+        xmlFree(xmlPropValue);
+        xmlPropValue = NULL;
+        
+        if (prefValue < 0 || prefValue > 100) 
+        {
+            dwError = ERROR_TDNF_MLK_PARSER_MISSING_URL_ATTR;
+            pr_err("XML Parser Warning: Bad value (\"%s\") of \"preference\""
+                   "attribute in url element (should be in range 0-100)", value);
+            BAIL_ON_TDNF_ERROR(dwError);
+        } 
+        else 
+        {
+            ml_url_info->preference = prefValue;
+        }
+        value = NULL;
+    }
+
+    //Get URL Content.
+    xmlContValue = xmlNodeGetContent(node);
+    if(xmlContValue == NULL)
+    {
+        dwError = ERROR_TDNF_MLK_PARSER_MISSING_URL_CONTENT;
+        pr_err("URL is no present in URL element", value);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    value = (const char*)xmlContValue;
+    ml_url_info->url = strdup(value);
```

Comment:
I see too many usage of strdup, cant we pass the pointer by reference and let the caller free it?

Reply:
used the TDNF function.